File: ts/1402.ts
function maxSatisfaction(satisfaction: number[]): number {
    satisfaction.sort((a, b) => a - b)
    let dp: number[][] = new Array(satisfaction.length + 1)
        .fill(0)
        .map( () => new Array(satisfaction.length + 1).fill(0) )

    let max = 0
    for (let i = 1; i <= satisfaction.length; i++) {
        for (let j = 1; j < i; j++) {
            dp[i][j] = Math.max(dp[i-1][j], dp[i-1][j-1] + satisfaction[i-1] * j)
            max = Math.max(max, dp[i][j])
        }

        dp[i][i] = dp[i-1][i-1] + satisfaction[i-1] * i
        max = Math.max(max, dp[i][i])
    }

    // console.log(satisfaction)
    // console.log(dp)

    return max
};

// maxSatisfaction([-1,-8,0,5,-7])

File: kt/1402.kt
class Solution {
    fun maxSatisfaction(satisfaction: IntArray): Int {
        satisfaction.sort()
        var sum = 0
        var start = 0
        for (i in satisfaction.indices.reversed()) {
            if (sum + satisfaction[i] > 0) {
                sum += satisfaction[i]
            } else {
                start = i + 1
                break
            }
        }


        // println(satisfaction)
        // println(start)

        return satisfaction
            .filterIndexed {idx, _ -> idx >= start}
            .mapIndexed { index, i -> (index + 1) * i }
            .sum()
    }
}

// fun main(args: Array<String>) {
//     println(Solution().maxSatisfaction(intArrayOf(-1,-8,0,5,-7)))
// }